grabber/main.go

```diff
@@ -142,72 +142,96 @@ func getLatestBlockNumber(importer *backend.Backend) int64 {
 	}
 	return number
 }
+
+// backfill continuously loops over all blocks in reverse order, verifying that all
+// data is loaded and consistent, and reloading as necessary.
 func backfill(importer *backend.Backend, startFrom int64) {
 	blockNumber := startFrom
 	if startFrom <= 0 {
 		blockNumber = getLatestBlockNumber(importer)
 	}
+	var hash common.Hash // Expected hash.
 	for {
+		logger := log.With().Int64("blockNumber", blockNumber).Logger()
 		if (blockNumber % 1000) == 0 {
-			log.Info().Int64("Block", blockNumber).Msg("Checking block in backfill")
+			logger.Info().Msg("Backfill progress")
 		}
-		blocksFromDB := importer.GetBlockByNumber(blockNumber)
-		if blocksFromDB == nil {
-			log.Info().Int64("Backfilling the block:", blockNumber).Msg("Getting block in backfill")
-			block, err := importer.BlockByNumber(blockNumber)
-			if block != nil {
-				importer.ImportBlock(block)
+		var backfill bool
+		dbBlock := importer.GetBlockByNumber(blockNumber)
+		if dbBlock == nil {
+			// Missing.
+			backfill = true
+			logger = logger.With().Str("reason", "missing").Logger()
+		} else if !common.EmptyHash(hash) && dbBlock.BlockHash != hash.Hex() {
+			// Mismatch with expected hash from parent of previous block.
+			backfill = true
+			logger = logger.With().Str("reason", "hash").Logger()
+		} else {
+			// Note parent as next expected hash.
+			hash = common.HexToHash(dbBlock.ParentHash)
+		}
+		if backfill {
+			logger.Info().Msg("Backfilling block")
+			rpcBlock, err := importer.BlockByNumber(blockNumber)
+			if rpcBlock != nil {
+				importer.ImportBlock(rpcBlock)
 				if err != nil {
-					log.Fatal().Err(err).Msg("importBlock - backfill")
+					logger.Fatal().Err(err).Msg("importBlock - backfill")
 				}
+				// Note parent as next expected hash.
+				hash = rpcBlock.ParentHash()
+			} else {
+				hash = common.Hash{}
 			}
+		} else {
+			hash = checkTransactionsConsistency(importer, blockNumber)
 		}
-		checkParentForBlock(importer, blockNumber, 5)
-		checkTransactionsConsistency(importer, blockNumber)
 		if blockNumber > 0 {
 			blockNumber = blockNumber - 1
 		} else {
 			blockNumber = getLatestBlockNumber(importer)
+			// No expected hash for latest.
+			hash = common.Hash{}
 		}
 	}
 }
 
-func checkParentForBlock(importer *backend.Backend, blockNumber int64, numBlocksToCheck int) {
-	numBlocksToCheck--
+// checkAncestors scans the ancestors of a block to verify that they exists and their hashes match.
+func checkAncestors(importer *backend.Backend, blockNumber int64, ancestorsToCheck int) {
```

Comment:
I found the recursion a little difficult to read (and the error case could infinitely retry/recurse!), so I rewrote it in to a loop, and changed the error to be fatal, which is in line with how it is handled elsewhere (retry happens internally).